Fix part of #19435: Topic editor migration

hrshkshri opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #19435 
2. Migrates Topic Editor page.

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [x] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)

## Proof of change


https://github.com/oppia/oppia/assets/108923011/72c1c59f-3d8d-4b1f-ab41-084b47878c1c

## Responsive

### oppiatestserver (Before )

https://github.com/oppia/oppia/assets/108923011/2922bb81-122a-40f0-8259-dad700746fa9

### localserver


https://github.com/oppia/oppia/assets/108923011/44bf5b24-1303-4a7c-9caf-512a294ff112



[Debugging doc](https://docs.google.com/document/d/1huNR9U8f0Xb0zRD8_C0HIbP2qbOHFOVigFY5Lqx75Jg/edit?usp=sharing) 

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

hrshkshri committed:
changes as per issue description

hrshkshri committed:
access gaurd to anular service

hrshkshri committed:
updated topiceditor module

hrshkshri committed:
updated topiceditor module

hrshkshri committed:
updated topiceditor module

hrshkshri committed:
change directive to component

DubeySandeep commented:
@hrshkshri Have you tried adding ModalCOmponent in the module declaration and entrypoint?

hrshkshri commented:
> @hrshkshri Have you tried adding ModalCOmponent in the module declaration and entrypoint?

`CreateNewStoryModalComponent` is already added in `topic-editor-tab.module.ts` declaration and entrypoint.

Also, in addition to that i thought this might be an error because `shared-component.module.ts` is in another folder/module and `shared-component.component.ts` is responsible for opening the modal (backend services), so, i added `CreateNewStoryModalComponent` in `shared-component.module.ts` entrypoint and declaration too.

DubeySandeep commented:
@hrshkshri Do we need to add these ModalComponents inside shared.module?

DubeySandeep commented:
> I thought the shared-component.module is also responsible for backend service, therefore I just tried to add modalComponent in shared-component.module.ts too

Are you still facing the issue?

hrshkshri commented:
> > I thought the shared-component.module is also responsible for backend service, therefore I just tried to add modalComponent in shared-component.module.ts too
> 
> Are you still facing the issue?

Yes @DubeySandeep 

Also, saw this file change in another migration PR https://github.com/oppia/oppia/pull/19745/files#diff-be3eee9da83a1135b7c6a3f34fbcc52be9b00b4e43c15b9c796c5c50c66a7640

In this PR module.ts also have `Provider` which provide a modalService 

Do i need to implement any logic related to that (still thinking of that approach)

hrshkshri committed:
fixed modalComponent

hrshkshri commented:
> > > I thought the shared-component.module is also responsible for backend service, therefore I just tried to add modalComponent in shared-component.module.ts too
> > 
> > 
> > Are you still facing the issue?
> 
> Yes @DubeySandeep
> 
> Also, saw this file change in another migration PR https://github.com/oppia/oppia/pull/19745/files#diff-be3eee9da83a1135b7c6a3f34fbcc52be9b00b4e43c15b9c796c5c50c66a7640
> 
> In this PR module.ts also have `Provider` which provide a modalService
> 
> Do i need to implement any logic related to that (still thinking of that approach)

UPDATE: Fixed the problem ( Indeed needed the above approach )

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

hrshkshri committed:
backend test

hrshkshri committed:
lint

hrshkshri committed:
updated backend test

hrshkshri committed:
backend test

hrshkshri committed:
test permission

hrshkshri committed:
backend service and specs test

hrshkshri committed:
updated backedTest

hrshkshri commented:
Hi @DubeySandeep 
I am having some trouble with the backend test (more specifically with `URL`).

after migration, the the URL would be `http://localhost/access_validation_handler/can_access_topic_editor/<topic_id>`

I am constantly getting `error 404 could not find the page`

DubeySandeep commented on core/controllers/access_validators_test.py:
```diff
@@ -416,3 +418,59 @@ def test_validation_returns_false_if_given_user_is_non_existent(
+        self
+    ) -> None:
+        self.login(self.CURRICULUM_ADMIN_EMAIL)
+        self.get_html_response(
```
Why are we looking for HTML response here?

hrshkshri replied:
Do we need to look at json response?
getting same error with json response

DubeySandeep replied:
can you check wether moving this to get_json resolves the issue?

hrshkshri replied:
just did in my local got this error again

```
=====================================================================
ERROR: test_access_topic_editor_page_with_curriculum_admin (core.controllers.access_validators_test.TopicEditorPageAccessValidationHandlerTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/harshkeshari/opensource/oppia/core/controllers/access_validators_test.py", line 463, in test_access_topic_editor_page_with_curriculum_admin
    self.get_json(
  File "/Users/harshkeshari/opensource/oppia/core/tests/test_utils.py", line 2885, in get_json
    json_response = self.testapp.get(
  File "/Users/harshkeshari/opensource/.direnv/python-3.8/lib/python3.8/site-packages/webtest/app.py", line 324, in get
    return self.do_request(req, status=status,
  File "/Users/harshkeshari/opensource/.direnv/python-3.8/lib/python3.8/site-packages/webtest/app.py", line 646, in do_request
    self._check_status(status, res)
  File "/Users/harshkeshari/opensource/.direnv/python-3.8/lib/python3.8/site-packages/webtest/app.py", line 680, in _check_status
    raise AppError(
webtest.app.AppError: Bad response: 404 Not Found (not 200)
)]}'
{"error": "Could not find the page http://localhost/access_validation_handler/can_access_topic_editor/7EMjW9QmKqzm.", "status_code": 404}

---------------------------------------------------------------------
FAILED    core.controllers.access_validators_test.TopicEditorPageAccessValidationHandlerTests: 1 errors, 0 failures

Ran 4 tests in 1 test class.
(1 ERRORS, 0 FAILURES)
Traceback (most recent call last):
  File "/Users/harshkeshari/.pyenv/versions/3.8.15/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/harshkeshari/.pyenv/versions/3.8.15/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/harshkeshari/opensource/oppia/scripts/run_backend_tests.py", line 683, in <module>
    main()
  File "/Users/harshkeshari/opensource/oppia/scripts/run_backend_tests.py", line 592, in main
    raise Exception(
Exception: 1 errors, 0 failures
```

seanlip replied:
Your URL looks weird. Shouldn't it be localhost:some-port? What do other backend tests for similar functionality look like and what are the differences between those tests and this one?

hrshkshri replied:
Hi @seanlip 
the URL is correct, I intentionally disturbed the test of `BlogAuthorProfilePageAccessValidationHandlerTests` just to check if my URL followed the same pattern as other tests.
![Screenshot 2024-02-20 at 9 36 22 AM](https://github.com/oppia/oppia/assets/108923011/28d51788-87b2-419d-b468-6f048e7b857d)

and got the error 
![Screenshot 2024-02-20 at 9 36 33 AM](https://github.com/oppia/oppia/assets/108923011/953d356c-4097-4f34-9d78-15eee72e10e7)

DubeySandeep commented on core/controllers/access_validators_test.py:
```diff
@@ -416,3 +418,59 @@ def test_validation_returns_false_if_given_user_is_non_existent(
+        self
+    ) -> None:
+        self.login(self.NEW_USER_EMAIL)
+        self.get_html_response(
```
Ditto

hrshkshri replied:
@DubeySandeep  Very Sorry for the late response,
The reason for using `get_html_response` here  is that the test expects an HTML response, using `get_json` leads to an assertion error (vice-versa is also true.) something like this
```
Traceback (most recent call last):
  File "/Users/harshkeshari/opensource/oppia/core/controllers/access_validators_test.py", line 551, in test_cannot_access_topic_editor_page_with_invalid_topic_id
    self.get_json(
  File "/Users/harshkeshari/opensource/oppia/core/tests/test_utils.py", line 2985, in get_json
    return self._parse_json_response(
  File "/Users/harshkeshari/opensource/oppia/core/tests/test_utils.py", line 2937, in _parse_json_response
    self.assertEqual(json_response.content_type, 'application/json')
AssertionError: 'text/html' != 'application/json'
- text/html
+ application/json
```

DubeySandeep commented on core/controllers/access_validators.py:
```diff
@@ -305,3 +306,41 @@ def get(self, author_username: str) -> None:
+        if topic is None:
+            raise self.PageNotFoundException(
+                Exception('The topic with the given id doesn\'t exist.'))
+        pass
```
why do we need pass here?

hrshkshri replied:
because we removed ` self.render_template('topic-editor-page.mainpage.html')` from previous class "TopicEditorPage" at topic_editor.py

DubeySandeep replied:
So you can just remove that line! (See other handle for reference)

hrshkshri replied:
Done

DubeySandeep commented on core/controllers/access_validators_test.py:
```diff
@@ -416,3 +418,59 @@ def test_validation_returns_false_if_given_user_is_non_existent(
+        self.add_user_role(
+            self.CURRICULUM_ADMIN_USERNAME, feconf.ROLE_ID_CURRICULUM_ADMIN)
+
+        self.topic_id = topic_fetchers.get_new_topic_id()
```
It looks like you are just fetching a new ID but not creating a new topic because this `if topic is None:` is working and `PageNotFoundException` is raised (404)!

DubeySandeep replied:
@hrshkshri I think this will resolve the 404 issue!

hrshkshri replied:
Hmmm, right.

hrshkshri committed:
updated backend tests

hrshkshri committed:
fixed api service url

hrshkshri committed:
frontend tests

hrshkshri committed:
lint

hrshkshri committed:
coverage test

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

hrshkshri committed:
update access_validators

hrshkshri committed:
fixed url

hrshkshri commented:
Hi @DubeySandeep @seanlip @kevintab95 PTAL

I am not sure why the TS test is failing, which is not related to the changes, I came across this comment https://github.com/oppia/oppia/issues/19217#issuecomment-1949068833 related to the same failing test, so I also cross-checked TS file changes, still, the test is failing.

hrshkshri commented on core/controllers/access_validators.py:
```diff
@@ -305,3 +306,41 @@ def get(self, author_username: str) -> None:
+        topic = topic_fetchers.get_topic_by_id(topic_id, strict=False)
+
+        if topic is None:
+            raise self.PageNotFoundException(
```
The backend coverage test is failing on this line.

hrshkshri commented on core/controllers/access_validators.py:
```diff
@@ -305,3 +306,41 @@ def get(self, author_username: str) -> None:
+        topic = topic_fetchers.get_topic_by_id(topic_id, strict=False)
+
+        if topic is None:
+            raise self.PageNotFoundException(
```
@DubeySandeep, The backend coverage test is failing for this particular line,
![Screenshot 2024-02-19 at 1 29 34 AM](https://github.com/oppia/oppia/assets/108923011/d528fe65-c2c2-47c6-91ac-a208195f73ce)


I noticed, the above handlers with `raise self.PageNotFoundException( ` line is also showing failed coverage, such as line 306, 260 ... etc

DubeySandeep replied:
Do we have a test where the user trying to access the handler with an invalid topic_id?

hrshkshri replied:
Yes we do.

seanlip replied:
Then why is the test failing? Is this line executed by the test in question (which test is it)?

You probably need to do a bit of debugging here.

hrshkshri replied:
This test is fixed now.
Thanks

hrshkshri commented on core/controllers/access_validators_test.py:
```diff
@@ -416,3 +418,72 @@ def test_validation_returns_false_if_given_user_is_non_existent(
+                self.topic_id), expected_status_int=200)
+        self.logout()
+
+    def test_cannot_access_topic_editor_page_with_invalid_topic_id(
```
@DubeySandeep 
here is the test for invalid_id

DubeySandeep replied:
This test is failing over here: https://github.com/oppia/oppia/blob/develop/core/controllers/acl_decorators.py#L3669

generate invalid_id using get_new_topic_id

hrshkshri replied:
Hi @DubeySandeep 

```
def test_cannot_access_topic_editor_page_with_invalid_topic_id(
        self
    ) -> None:
        self.login(self.NEW_USER_EMAIL)
        self.get_json(
            '%s/can_access_topic_editor/%s' % (
                ACCESS_VALIDATION_HANDLER_PREFIX,
                topic_fetchers.get_new_topic_id()), expected_status_int=401)
        self.logout()
```

changed `invalid_id` to `topic_fetchers.get_new_topic_id()` for valid id 
still coverage test is failing

![Screenshot 2024-02-19 at 2 26 10 PM](https://github.com/oppia/oppia/assets/108923011/b2c05575-a6e4-43ff-a0f3-069387256ad2)

seanlip replied:
This is quite obviously not hitting the line in question. It's raising a 401 and not a 404 (which would correspond to PageNotFoundException). You'll need to do some analysis -- e.g. perhaps this case is already caught by the ACL decorator?

hrshkshri replied:
Hmmm, right. thanks
updated the tests

hrshkshri committed:
updated invalid id test

seanlip reviewed (changes requested):
(Looked at codeowner files only)

hrshkshri committed:
updated invalid_id test

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

DubeySandeep commented:
@hrshkshri CI builds are failing on this PR, PTAL!

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

hrshkshri commented:
> @hrshkshri CI builds are failing on this PR, PTAL!

Hey @DubeySandeep 
I have been trying to fix these failed CI builds.
The "TypeScript" test is giving errors which is not related to my changes. as I mentioned here https://github.com/oppia/oppia/pull/19747#issuecomment-1951428799

I am curious about other lighthouse accessibility tests, tried the approach mentioned in the wiki ( https://github.com/oppia/oppia/wiki/Lighthouse-Tests ) and tried, and maybe I can have some help here. : )

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

nikitaevg commented on main.py:
```diff
@@ -278,6 +278,11 @@ def get_redirect_route(
         access_validators.ViewLearnerGroupPageAccessValidationHandler),
 
+    get_redirect_route(
+        r'%s/can_access_topic_editor/<topic_id>' %
```
ACCESS_VALIDATION_HANDLER_PREFIX already mentions that this is about access validation, so having "can_access" is redundant in this name IMO. Maybe just "/access_validation_handler/topic_editor/<topic_id>"?

hrshkshri replied:
Hmmm. I'll consider it

hrshkshri replied:
Hey @nikitaevg After looking at all of the migration PRs, I think `can_access_....` makes sense to me, since we are using a similar format for all access validators.

https://github.com/oppia/oppia/blob/e51de7ed92ecef9fa9850d9f552bb558bd401209/main.py#L233-L286

hrshkshri committed:
lint

DubeySandeep commented:
@hrshkshri Lnt checks are failing on this PR, can you confirm wether pre-push hooks are working fine locally i.e, lint checks are running after `git push <feature-branch>` command?

hrshkshri commented:
Sorry about the lint checks.

I resolved the merge conflict on GitHub itself and there was one extra
space left.

The main concern is lighthouse test

And teah pre-push hooks are fine


On Sat, 24 Feb, 2024, 3:31 am Sandeep Dubey, ***@***.***>
wrote:

> @hrshkshri <https://github.com/hrshkshri> Lnt checks are failing on this
> PR, can you confirm wether pre-push hooks are working fine locally i.e,
> lint checks are running after git push <feature-branch> command?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/19747#issuecomment-1962059881>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZ7ARA6QMUOWE5BRYS35WYLYVEGVPAVCNFSM6AAAAABDH3MJP6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRSGA2TSOBYGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hrshkshri commented:
> @hrshkshri Lnt checks are failing on this PR, can you confirm wether pre-push hooks are working fine locally i.e, lint checks are running after `git push <feature-branch>` command?

Sorry about the lint checks.

I resolved the merge conflict on GitHub itself and there was one extra
space left.

The main concern is the lighthouse test

And yeah pre-push hooks are fine

DubeySandeep commented:
> I resolved the merge conflict on GitHub itself and there was one extra

Generally, we don't suggest resolving conflicts on Github (for the same reason i.e, git hooks), please avoid doing so as it adds load to our build queue.

> The main concern is the lighthouse test


I've request permission on your debugging doc, will be able to help there once I get access.

hrshkshri commented:
> Generally, we don't suggest resolving conflicts on Github (for the same reason i.e, git hooks), please avoid doing so as it adds load to our build queue.
 
 Ohk, will take care of that.
 
> I've request permission on your debugging doc, will be able to help there once I get access.

I gave the permission 
thanks

hrshkshri committed:
added previous test

hrshkshri committed:
revert

DubeySandeep commented:
@hrshkshri is this PR ready for review?

hrshkshri commented:
> @hrshkshri is this PR ready for review?

Still working on the lighthouse test.

hrshkshri committed:
revert ts file

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

hrshkshri committed:
backend test

hrshkshri committed:
removed 404 test

hrshkshri committed:
backend coverage

hrshkshri committed:
changed routing guard

hrshkshri committed:
changed rights

hrshkshri committed:
changed rights

hrshkshri committed:
revert TS file

hrshkshri committed:
revert TS file

DubeySandeep commented:
@hrshkshri, could you take a look at the comments under the debugging document? I'm available to assist with resolving the problem. I believe I added a crucial note that might contain a hint on possible suspect for the issue.

hrshkshri commented:
> @hrshkshri, could you take a look at the comments under the debugging document? I'm available to assist with resolving the problem. I believe I added a crucial note that might contain a hint on possible suspects for the issue.

Hi @DubeySandeep 
I took a pass in the debugging doc, while we are solving this particular problem can I get assigned to another page for migration ( maybe a simpler one, just wanted to understand the working )

DubeySandeep commented on core/templates/pages/topic-editor-page/topic-editor-page.mainpage.html:
```diff
@@ -1,37 +0,0 @@
-
-    </oppia-root>
-    @load('pages/footer_js_libs.html')
-    <script src="/third_party/static/MathJax-2.7.5/MathJax.js?config=default"></script>
```
@hrshkshri  Do we need to add this in the root component?

hrshkshri replied:
Yeah, I believe we must add this.
Thanks.

hrshkshri replied:
Done

DubeySandeep commented:
> I took a pass in the debugging doc, while we are solving this particular problem can I get assigned to another page for migration ( maybe a simpler one, just wanted to understand the working )

Sure, I'll check if I can find something simpler though this PR is close to merge. Anyway, we can come back after solving another similar task.

hrshkshri committed:
added script

hrshkshri commented:
> > I took a pass in the debugging doc, while we are solving this particular problem can I get assigned to another page for migration ( maybe a simpler one, just wanted to understand the working )
> 
> Sure, I'll check if I can find something simpler though this PR is close to merge. Anyway, we can come back after solving another similar task.

Thanks

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

hrshkshri commented:
Hey @DubeySandeep 
Just to check if we can start review on this PR

Thanks

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

hrshkshri committed:
prettier

hrshkshri committed:
prettier

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

seanlip commented:
@hrshkshri Please check the conversation history on https://github.com/oppia/oppia/pull/19747 and make sure you have responded to all comments before requesting review. Thanks.

hrshkshri commented:
> @hrshkshri Please check the conversation history on #19747 and make sure you have responded to and addressed all comments (including earlier ones) before requesting a follow-up review. Thanks.

Hey @seanlip 
I've responded and addressed all of the previous conversation history, not sure which one I am missing. 

Also, just want to add one I noticed that `lighthouse CI accessibility tests` are failing in this PR ( [log](https://github.com/oppia/oppia/actions/runs/8259169318/job/22592634688?pr=19747) ) which is common in  #19882  `lighthouse CI accessibility tests` is also failing in that PR too. ( [log](https://github.com/oppia/oppia/actions/runs/8167697524/job/22328519500) )

seanlip commented:
> I've responded and addressed all of the previous conversation history, not sure which one I am missing.

@hrshkshri This one originally, which you responded to 2 days ago (https://github.com/oppia/oppia/pull/19747#discussion_r1494903291) when you posted the "not sure which one I am missing" comment, but also this one: https://github.com/oppia/oppia/pull/19747#discussion_r1493329654

seanlip committed:
Merge branch 'develop' into topic-editor-migration

seanlip commented:
I'm closing and reopening this to see if it helps with the CI issues.

seanlip commented:
This also has no response: https://github.com/oppia/oppia/pull/19747#discussion_r1501071415. It just says "I'll consider it" but it doesn't say what decision has been taken.

seanlip commented on core/controllers/access_validators_test.py:
```diff
@@ -469,3 +471,82 @@ def test_validation_returns_false_if_given_user_is_non_existent(
+        self.login(self.VIEWER_EMAIL)
+        self.get_json(
+            '%s/can_access_topic_editor/%s' % (
+            ACCESS_VALIDATION_HANDLER_PREFIX,
```
Indent this and the next line by 4 since you are creating a new scope with the '('.

hrshkshri replied:
Done

seanlip commented on core/controllers/access_validators_test.py:
```diff
@@ -469,3 +471,82 @@ def test_validation_returns_false_if_given_user_is_non_existent(
+        self.get_json(
+            '%s/can_access_topic_editor/%s' % (
+            ACCESS_VALIDATION_HANDLER_PREFIX,
+            self.topic_id), expected_status_int=401)
```
Break before ')' and make sure the line with ')' has the same indentation as the line with the corresponding '('

hrshkshri replied:
Done.
Also, I have used similar indentation throughout the test, have a look.
Thanks

seanlip commented on core/controllers/access_validators_test.py:
```diff
@@ -469,3 +471,82 @@ def test_validation_returns_false_if_given_user_is_non_existent(
+            ACCESS_VALIDATION_HANDLER_PREFIX,
+            self.topic_id), expected_status_int=401)
+
+    def test_access_topic_editor_page_with_guest_user(self) -> None:
```
You're logging in for this test, so how can these be "guest users"? Guests typically refer to logged-out users.

hrshkshri replied:
Ahh, Sorry, updated the test.

seanlip replied:
I don't understand, what updates have you made? The test is still named "guest user" and its implementation still includes logging in.

hrshkshri replied:
Removed the logging in implementation

seanlip commented on core/controllers/access_validators_test.py:
```diff
@@ -469,3 +471,82 @@ def test_validation_returns_false_if_given_user_is_non_existent(
+    def test_access_topic_editor_page_without_logging_in(self) -> None:
+        self.get_json(
+            '%s/can_access_topic_editor/%s' % (
+            ACCESS_VALIDATION_HANDLER_PREFIX,
```
See indentation notes below.

hrshkshri replied:
Updated

seanlip commented on core/controllers/access_validators_test.py:
```diff
@@ -469,3 +471,82 @@ def test_validation_returns_false_if_given_user_is_non_existent(
         self.logout()
+
+
+class TopicEditorPageAccessValidationHandlerTests(test_utils.EmailTestBase):
```
Why is this using EmailTestBase and not the generic one?

hrshkshri replied:
The test i removed from the `topic_editor_test` file was using the `EmailTestBase` that's why i used it here too, Updated it to `GenericTestBase`

seanlip commented on core/controllers/access_validators_test.py:
```diff
@@ -469,3 +471,82 @@ def test_validation_returns_false_if_given_user_is_non_existent(
+        self.get_html_response(
+            '%s/can_access_topic_editor/%s' % (
+                ACCESS_VALIDATION_HANDLER_PREFIX,
+                self.topic_id), expected_status_int=200)
```
Shouldn't you check what the response actually is, or assert that some expected string is contained in the response?

hrshkshri replied:
I am following the structure just like other migrated `access_validator_tests` right above.

Also, Do we need to check, if we use the `get_json` response but the expected response is in `HTML` it gives an error and will be solved if we use `get_html_response`.

seanlip commented on core/controllers/access_validators_test.py:
```diff
@@ -469,3 +471,82 @@ def test_validation_returns_false_if_given_user_is_non_existent(
+        self.get_html_response(
+            '%s/can_access_topic_editor?topic_id=invalid_id' % (
+                ACCESS_VALIDATION_HANDLER_PREFIX,
+                ), expected_status_int=404)
```
Deindent by 4. The general rule is that closing parens should be on the same indentation level as the line with the corresponding opening paren.

hrshkshri replied:
Done

seanlip reviewed (changes requested):
Re-reviewed my codeowner files, PTAL.

(But also please see the previous notes about responding to all previous comments.)

seanlip commented:
@hrshkshri Also a note about the lighthouse test failure -- if they are failing in develop then that might be a reason for them to fail here, but if they are failing here and on other PRs only then I think you might need to look into the cause. It might be worth talking to @srijanreddy98 and @DubeySandeep about this too since it seems to be affecting at least two Angular migration PRs, or perhaps you can discuss with @shivanandan17 to see if you can both figure out a solution.

But I think this is a problem whose cause needs to be looked at. I don't think restarting is going to fix it.

hrshkshri committed:
indent and guest test

hrshkshri committed:
updated genericTestBase

hrshkshri committed:
lint

hrshkshri commented:
> @hrshkshri Also a note about the lighthouse test failure -- if they are failing in develop then that might be a reason for them to fail here, but if they are failing here and on other PRs only then I think you might need to look into the cause. It might be worth talking to @srijanreddy98 and @DubeySandeep about this too since it seems to be affecting at least two Angular migration PRs, or perhaps you can discuss with @shivanandan17 to see if you can both figure out a solution.
> 
> But I think this is a problem whose cause needs to be looked at. I don't think restarting is going to fix it.

Hey @seanlip
Yes, restarting the test did not help at all,
Also, IDK if this will make any sense but I noticed it is only failing for migration of an editor page eg, `[Skill-Editor-Page](https://github.com/oppia/oppia/pull/19882)`  and here at `Topic-Editor-Page`, I am waiting for @srijanreddy98 to finish the `[Exploration-Editor-Page](https://github.com/oppia/oppia/pull/19889)` migration just to see if lighthouse CI fails there too.
We haven't migrated any editor-page so far, (maybe i am wrong here ) : )

Hmm, the story-editor-page ( here-- https://github.com/oppia/oppia/issues/19435#issue-2072988628 ) is ready for migration should I take a look ...?

Also, I have addressed all of the previous comments.
Thanks

hrshkshri commented:
> This also has no response: [#19747 (comment)](https://github.com/oppia/oppia/pull/19747#discussion_r1501071415). It just says "I'll consider it" but it doesn't say what decision has been taken.

replied to this.

hrshkshri commented:
> > I've responded and addressed all of the previous conversation history, not sure which one I am missing.
> 
> @hrshkshri This one originally, which you responded to 2 days ago ([#19747 (comment)](https://github.com/oppia/oppia/pull/19747#discussion_r1494903291)) when you posted the "not sure which one I am missing" comment, but also this one: [#19747 (comment)](https://github.com/oppia/oppia/pull/19747#discussion_r1493329654)

Sorry, replied now.

seanlip commented on core/controllers/access_validators_test.py:
```diff
@@ -469,3 +471,86 @@ def test_validation_returns_false_if_given_user_is_non_existent(
+        self.get_html_response(
+            '%s/can_access_topic_editor/%s' % (
+                ACCESS_VALIDATION_HANDLER_PREFIX, self.topic_id
+            ), expected_status_int=200
```
Re: https://github.com/oppia/oppia/pull/19747#discussion_r1527218260 (Shouldn't you check what the response actually is, or assert that some expected string is contained in the response?) -- I suggest still checking something here. I looked at the code above and it seems like it doesn't follow indentation guidelines either so I am not sure if this was missed in previous reviews as well. Is there anything that would help identify the response as the right one (what does the response look like)?

> Also, Do we need to check, if we use the get_json response but the expected response is in HTML it gives an error and will be solved if we use get_html_response.

I don't really understand your question, sorry. What do you mean by "if we use the get_json response" and what is the actual scenario here?

hrshkshri replied:
Sorry @seanlip, about my previous comment (It was not clear what I wanted to say )

What I understand so far is that for every request at `URL` there is a specific content type ( either `application/json`, `text/html` )

I check nearly all of the merged migration PR, and all of them directly use `get_html_response ` or `get_json`

Now here is what going on...

for `get_html_response` 
* checks for expected_errors https://github.com/oppia/oppia/blob/001e4c464289cefeac36445a7adfc05076316e07/core/tests/test_utils.py#L2806
https://github.com/oppia/oppia/blob/001e4c464289cefeac36445a7adfc05076316e07/core/tests/test_utils.py#L2832
* checks what the type of response we are getting from the URL ( either `application/json`, `text/html` )
https://github.com/oppia/oppia/blob/001e4c464289cefeac36445a7adfc05076316e07/core/tests/test_utils.py#L2834

which means if we use `get_json` instead of `get_html_response`  it will give an assertion error based on the above checks
* For the first check
```
    self.assertTrue(response.status_int >= 400)
AssertionError: False is not true
```
* for the second check
```
AssertionError: 'application/json' != 'text/html'
- application/json
+ text/html
```

Same for the `get_json`

https://github.com/oppia/oppia/blob/001e4c464289cefeac36445a7adfc05076316e07/core/tests/test_utils.py#L2982-L2984

https://github.com/oppia/oppia/blob/001e4c464289cefeac36445a7adfc05076316e07/core/tests/test_utils.py#L2936-L2938


So, we already have assertion checks in `get_json` and `get_html_response`based on what response we are getting from the URL.

Also, is it a good idea to first check the response type in access_handler_test for each request to URL and then use `get_json` or `get_html_response` ( knowing that both of these functions will already do the same job )?

seanlip replied:
Hi @hrshkshri, I still don't understand what you are trying to say, sorry.

I get that we are doing generic checks for HTML and JSON responses. The first paragraph of my comment talks about checks on the **contents** of the response, not generic checks on the format. Does that help resolve some of the confusion here?

I also still can't understand what your last question means exactly -- it might be better to phrase your question as "should I go with option 1 or option 2" for this sort of thing, because I'm not sure what alternative you are asking me to help compare against.

hrshkshri replied:
Hi @seanlip, Sorry for the late response...

> The first paragraph of my comment talks about checks on the contents of the response, not generic checks on the format. Does that help resolve some of the confusion here?

I understand the concern here, After checking all of the PRs ( Related to migration ), and the old `handler test` ( `...._test.py`[ /core/controllers ](https://github.com/oppia/oppia/tree/aa267c3e77b571cef9e6e7479119506f6651f55a/core/controllers )) which are related to the access validation route, what I am trying to say is that all of them follow the similar format ( At least what I understand )

Also, above I was trying to make a point that we shouldn't check if the response we are getting is either `application/json` or `text/html` since the `get_json()` and `get_html_response()` functions are doing that job.

But as you mentioned earlier (https://github.com/oppia/oppia/pull/19747#discussion_r1527218260) about checking on the contents of the response (not sure how to do that for now) could be the discussion here, also not to forget we are still migrating the pages, therefore, https://github.com/oppia/oppia/issues/19435 needs to be updated for the same (If we are planning to do that so).

WDYT?

Thanks

seanlip replied:
Isn't it just checking `response.body`? There are some tests for that in the codebase. I think it's just adding a single line.

It's a reasonable question about whether we want to do that for all the tests. I'll defer to @DubeySandeep and @srijanreddy98 for that.

I still don't understand why the question of verifying json vs html came up, btw. Was it from my comment "what the response actually is"? If so, by that phrase, I meant the response content, not its type -- i.e. the same thing as I mentioned in the first paragraph of this response. Sorry for any confusion there.

hrshkshri replied:
>I still don't understand why the question of verifying json vs html came up, btw. Was it from my comment "what the response actually is"? If so, by that phrase, I meant the response content, not its type -- i.e. the same thing as I mentioned in the first paragraph of this response. Sorry for any confusion there.

Sorry, I got a little confused by the comment on "response" and brought up the json vs html thing.

>It's a reasonable question about whether we want to do that for all the tests. I'll defer to @DubeySandeep and @srijanreddy98 for that.

/cc @DubeySandeep @srijanreddy98

seanlip reviewed (changes requested):
Hi @hrshkshri, thanks. You might want to contact @srijanreddy98 and @DubeySandeep directly about the lighthouse error. 

Feel free to have a go at migrating story-editor too if you like. Thanks!

Also I left a couple more inline comments, PTAL.

hrshkshri committed:
updated test for guest users

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

hrshkshri committed:
lint

hrshkshri committed:
lint

hrshkshri commented:
https://github.com/oppia/oppia/pull/19747#issuecomment-2002093117 https://github.com/oppia/oppia/pull/19747#issuecomment-2002607664

/cc @DubeySandeep @srijanreddy98 

Thanks

hrshkshri committed:
Merge branch 'develop' into topic-editor-migration

hrshkshri committed:
typo
